### DISCUSSION TITLE
Global variable as parameter to function

### DISCUSSION THREAD

**Discussion by fspinar** [jspsych/jsPsych#2046] [General]
Hi all,

I am trying to modify a stimulus in a trial by trial basis. The stimulus depends both on answers from subjects, and a probability distribution, which I am defining globally. Then, when I define the stimulus as a function, it seems like I cannot pass the distribution as a parameter to the function. For some reason, despite explicitly adding the variable names as parameters, I cannot access the distribution variable inside the function and I get the following errors :

**First argument to jsPsych.randomization.sampleWithReplacement() must be an array
Uncaught TypeError: Cannot read property 'length' of undefined**

 How can I solve this ? The pseudocode is something like that: 
 
 _var distribituon = []; // whatever it is
 
 var trial = {
 type: "image-keyboard-response",
            stimulus: function (data) {
 var result = jsPsych.randomization.sampleWithReplacement(distribution[i], 1)[0];
if(result){
return "image1.png"
}else{
"image2.png"
}
}_

Thank you and best regards.

**Comment by becky-gilbert** [comment]
Hi @fspinar, the first argument to the randomization function has to be an _array_, but you're indexing the distribution array there, which means that you're passing one of the array _elements_ as the first argument. So try removing the square brackets so that the first argument is just the `distribution` array:
```js
var result = jsPsych.randomization.sampleWithReplacement(distribution, 1)[0];
```
Does that work?

Also, I noticed that you've included a `data` argument in your stimulus function. When you use a function for the `stimulus` parameter, it actually doesn't take any arguments, so `data` will be undefined inside that function. If you need access to the participant's previous responses inside this function, then you can use one of the [jsPsych data functions](https://www.jspsych.org/core_library/jspsych-data/) instead.

**Reply by fspinar** [reply]
Thanks a lot for the response, @becky-gilbert 

I actually made a typo in the question, distribution[i] itself is an array, so the issue must be something else. 
But how can I pass a parameter to
 stimulus: function (data) {
 }
 function ? I think I definitely need to pass the distribution as a parameter. This is because I need to create it for once and use it for many trials without changing it in blocks. If I create and use it inside the function then it is evaluated and changed whenever the trial is run, and I don't want that. 
 So I should either pass distribution as a parameter, or, somehow create it inside the function but do not change it in every trial.
 
 Thanks a lot.

**Reply by becky-gilbert** [reply]
Hi @fspinar, oh ok, if `distribution[i]` is itself an array then you're right, the indexing isn't the problem. If `distribution` is a global variable then you should be able to access it inside your stimulus function - you can check this and the value of `i` by adding `console.log`s to your stimulus function:
```js
stimulus: function() {
  console.log('distribution: ', distribution);
  console.log('i: ', i);
  // etc.
}
```
Maybe the problem is indexing with `i`. Is the `i` variable an iterator for a for loop? If so, then typically this won't work, because the whole for loop will finish executing when the page first loads, but the stimulus function doesn't run until later on (right before each trial starts).

As far as using `distribution` as a global variable, this should work fine and IMO would be the easiest solution. You should be able to use that variable to create other (trial-specific) variables, without changing its value. Here's a little dummy example:
```js
// global variable that we'll use to create other variables, but will remain unchanged
var global_array = [500,600,700]; 

var trial = {
  type: 'html-keyboard-response',
  stimulus: function() {
    var sampled_value = jsPsych.randomization.shuffle(global_array,1)[0];
    var current_stimulus = 'the sampled value is '+sampled_value.toString();
    return current_stimulus;
  }
};
```
Or if the problem is that you also need a variable that _does_ change across trials, for instance to track performance, then you can do that in a separate global variable so that your `distribution` variable remains unchanged. Here's another dummy example:
```js
// global variable that will remain unchanged
var global_static_array = [500,600,700]; 
// global variables that will change across trials
var global_stim_array = [];
var global_response_array = [];

var trial = {
  type: 'html-keyboard-response',
  stimulus: function() {
    var sampled_value = jsPsych.randomization.shuffle(global_static_array,1)[0];
    global_stim_array.push(sampled_value); // modify global array
    var current_stimulus = 'the sampled value is '+sampled_value.toString();
    return current_stimulus;
  },
  on_finish: function(data) {
    global_response_array.push(data.response); // modify global array
  }
};
```
Does this help at all? Apologies if not, and feel free to explain more about what you're trying to do and/or include more of your code.

**Comment by fspinar** [comment]
Hi again @becky-gilbert 

Thanks a lot for your response. For now I just got rid of my loop and defined several parameters manually. But I think the problem is using variable i as the index of the loop. As you said, the loop index becomes inconsistent with the index of global_array while executing the function. Right now defining variable without for loop is OK, so I will go with that.

Thanks a lot.
